packages/itwin/tree-widget/src/components/trees/stateless/common/UseNodeHighlighting.tsx

```diff
@@ -0,0 +1,204 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { useCallback, useEffect, useRef } from "react";
+import { isPresentationHierarchyNode } from "@itwin/presentation-hierarchies-react";
+
+import type { PresentationHierarchyNode, PresentationTreeNode } from "@itwin/presentation-hierarchies-react";
+
+interface UseNodeHighlightingProps {
+  rootNodes: PresentationTreeNode[] | undefined;
+  textToHighlight?: string;
+  caseSensitive?: boolean;
+  activeMatchIndex?: number;
+  onHighlightChanged?: (activeMatchIndex: number, matches: number) => void;
+}
+
+interface HighlightedChunk {
+  start: number;
+  end: number;
+}
+
+interface NodeChunkInfo {
+  chunks: HighlightedChunk[];
+  activeChunkIndex?: number;
+}
+
+interface NodeHighlightInfo {
+  startIndex: number;
+  matches: HighlightedChunk[];
+}
+
+interface HighlightState {
+  nodeInfoMap: Map<string, NodeHighlightInfo>;
+  totalMatches: number;
+}
+
+interface UseNodeHighlightingResult {
+  getLabel: (node: PresentationHierarchyNode) => React.ReactElement;
+}
+
+/** @internal */
+export function useNodeHighlighting({
+  rootNodes,
+  textToHighlight,
+  caseSensitive,
+  activeMatchIndex,
+  onHighlightChanged,
+}: UseNodeHighlightingProps): UseNodeHighlightingResult {
+  const state = useRef<HighlightState>({ nodeInfoMap: new Map(), totalMatches: 0 });
+  const activeMatchIndexRef = useRef(activeMatchIndex);
+
+  useEffect(() => {
+    activeMatchIndexRef.current = activeMatchIndex;
+  }, [activeMatchIndex]);
+
+  const getNodeChunkInfo = useCallback((nodeId: string, activeIndex?: number): NodeChunkInfo | undefined => {
+    const info = state.current.nodeInfoMap.get(nodeId);
+    if (!info) {
+      return undefined;
+    }
+    if (activeIndex === undefined) {
+      return { chunks: info.matches };
+    }
+    const isActive = info && activeIndex >= info.startIndex && activeIndex < info.startIndex + info.matches.length;
+    return isActive ? { activeChunkIndex: activeIndex - info.startIndex, chunks: info.matches } : { chunks: info.matches };
+  }, []);
+
+  const computeHighlightState = useCallback(() => {
```

Comment:
I think this can be extracted into separate function istead of using `useCallback`.

Reply:
Extracted functions where applicable.